### PR TITLE
#22: EDML: Made behaviour of toVarchar mapping if not string configurable

### LINKED ISSUE #22: EDML: define behaviour for toVarchar mapping if can not convert
At the moment the VS throws an exception if a property contains a value that can not be converted to the defined destination type.

Solution:
* Behaviour can be defined using edml
  Option:
   * ABORT
   * NULL
   * TO_STRING_IF_FAIL_ABORT (current defualt)
   * TO_STRING_IF_FAIL_NULL

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #22

### CODE REVIEW

File: doc/changes/changes_2.0.0.md [hunk 1]
```diff
@@ -9,11 +9,17 @@ Breaking API Changes:
...
 
 
 ## Features / Enhancements
 
 * #23: Refactored UDF factory API
+* #22: EDML: Made behaviour for toVarchar mapping if not string configurable
+
+## Bug Fixes
+
+* #24 Removed hardcoded UDF schema (#26)
```
Comment (jakobbraun) [inline doc/changes/changes_2.0.0.md:22]: i forgot to add this line in the last PR...

File: src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java [hunk 2]
```diff
@@ -53,4 +87,27 @@ private String handleOverflow(final String tooLongSourceString) {
...
 
+    /**
+     * This class is used to pass the result of {@link #mapStringValue(DocumentNode)} from the concrete implementation
+     * to this abstract class.
+     * 
+     * @implNote public so that it is accessible from test code
+     */
+    public static class Result {
+        private final String value;
+        private final boolean wasConverted;
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java:98]:
```suggestion
        private final boolean isConverted;
```

File: src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java [hunk 3]
```diff
@@ -53,4 +87,27 @@ private String handleOverflow(final String tooLongSourceString) {
         }
     }
 
+    /**
+     * This class is used to pass the result of {@link #mapStringValue(DocumentNode)} from the concrete implementation
+     * to this abstract class.
+     * 
+     * @implNote public so that it is accessible from test code
+     */
+    public static class Result {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java:96]: Result is too generic. How about, ParsedStringResult or MappedStringResult?

File: src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java [hunk 4]
```diff
@@ -25,15 +25,49 @@ public PropertyToVarcharColumnValueExtractor(final PropertyToVarcharColumnMappin
...
+        } else {
+            return handleConvertedResult(result);
+        }
+    }
+
+    private String handleConvertedResult(final Result stringValue) {
+        if (stringValue.wasConverted()
+                && this.column.getNotAStringBehaviour().equals(ConvertableMappingErrorBehaviour.ABORT)) {
+            throw new ColumnValueExtractorException(
+                    "An input value was not a string. This adapter could convert it to string, but you disabled this by setting notAStringBehaviour to ABORT.",
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java:49]:
```suggestion
                    "An input value is not a string. This adapter could convert it to string, but you disabled this by setting notAStringBehaviour to ABORT.",
```

File: src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java [hunk 5]
```diff
@@ -25,15 +25,49 @@ public PropertyToVarcharColumnValueExtractor(final PropertyToVarcharColumnMappin
...
+                    this.column);
+        } else if (stringValue.wasConverted()
+                && this.column.getNotAStringBehaviour().equals(ConvertableMappingErrorBehaviour.NULL)) {
+            return null;
+        } else {
+            return handleOverflowIfNecessary(stringValue.getValue());
+        }
+    }
+
+    private String handleCouldNotConvert() {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java:59]:
```suggestion
    private String handleNotConvertedResult() {
```

File: src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnMapping.java [hunk 6]
```diff
@@ -9,28 +9,33 @@
...
  */
 public final class PropertyToVarcharColumnMapping extends AbstractPropertyToColumnMapping {
-    private static final long serialVersionUID = 3465558198156097064L;//
+    private static final long serialVersionUID = -5267440896076306664L;//
     /** @serial */
     private final int varcharColumnSize;
     /** @serial */
     private final TruncateableMappingErrorBehaviour overflowBehaviour;
+    /** @serial */
+    private final ConvertableMappingErrorBehaviour notAStringBehaviour;
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnMapping.java:18]: How about naming this `nonAlphabeticBehaviour`? Or `nonStringBehaviour`?

**Review by morazow — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/document/mapping/reader/ColumnMappingReader.java [hunk 7]
```diff
@@ -146,5 +164,6 @@ private void abortIfAtRootLevel(final String mappingType, final boolean isRootLe
         private static final int DEFAULT_DECIMAL_SCALE = 0;
         private static final int DEFAULT_DECIMAL_PRECISION = 18;
         private static final String NOT_NUMERIC_BEHAVIOUR = "notNumericBehaviour";
+        private static final String NOT_A_STRING_BEHAVIOUR = "nonStringBehaviour";
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/reader/ColumnMappingReader.java:167]:
```suggestion
        private static final String NON_STRING_BEHAVIOUR = "nonStringBehaviour";
```

File: src/main/java/com/exasol/adapter/document/mapping/reader/ColumnMappingReader.java [hunk 8]
```diff
@@ -87,6 +88,21 @@ private MappingErrorBehaviour readMappingErrorBehaviour(final String key, final
         }
     }
 
+    private ConvertableMappingErrorBehaviour readConvertableMappingErrorBehaviour(final JsonObject definition) {
+        switch (definition.getString(EDML.NOT_A_STRING_BEHAVIOUR, "").toUpperCase()) {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/reader/ColumnMappingReader.java:92]: Change here also.

**Review by morazow — CHANGES_REQUESTED** [review]

**Review by morazow — APPROVED** [review]

**Pull request merged**